### PR TITLE
fix: correct osmo-multihop swapfee reduction

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #3346

### CODE REVIEW

File: x/gamm/keeper/multihop.go [hunk 1]
```diff
@@ -16,12 +16,55 @@ func (k Keeper) MultihopSwapExactAmountIn(
...
+			// determine the max swap fee
+			if swapFee.GT(maxSwapFee) {
+				maxSwapFee = swapFee
+			}
 		}
+	}
+
+	if additiveSwapFee.Quo(sdk.NewDec(2)).GT(maxSwapFee) {
+		maxSwapFee = additiveSwapFee.Quo(sdk.NewDec(2))
+	}
```
Comment (ValarDragon) [inline x/gamm/keeper/multihop.go:65]: We should move this to its own function

Comment (czarcas7ic) [reply x/gamm/keeper/multihop.go]: moved it to its own func, sadly cant have too much reuse between SwapOut and SwapIn due to having different route types

**Review by ValarDragon — APPROVED** [review]

File: x/gamm/keeper/multihop.go [hunk 2]
```diff
@@ -16,12 +17,29 @@ func (k Keeper) MultihopSwapExactAmountIn(
 	tokenIn sdk.Coin,
 	tokenOutMinAmount sdk.Int,
 ) (tokenOutAmount sdk.Int, err error) {
-	for i, route := range routes {
-		swapFeeMultiplier := sdk.OneDec()
-		if types.SwapAmountInRoutes(routes).IsOsmoRoutedMultihop() {
-			swapFeeMultiplier = types.MultihopSwapFeeMultiplierForOsmoPools.Clone()
+	isMultiHopRouted, routeSwapFee, sumOfSwapFees := false, sdk.Dec{}, sdk.Dec{}
+	route := types.SwapAmountInRoutes(routes)
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:21]:
```suggestion
	var (
		isMultiHopRouted bool
		routeSwapFee     sdk.Dec
		sumOfSwapFees    sdk.Dec
	)

	route := types.SwapAmountInRoutes(routes)
```

File: x/gamm/keeper/multihop.go [hunk 3]
```diff
@@ -16,12 +17,29 @@ func (k Keeper) MultihopSwapExactAmountIn(
...
+	if err := route.Validate(); err != nil {
+		return sdk.Int{}, err
+	}
+
+	// in this loop, we check if:
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:26]:
```suggestion
	// In this loop, we check if:
```

File: x/gamm/keeper/multihop.go [hunk 4]
```diff
@@ -16,12 +17,29 @@ func (k Keeper) MultihopSwapExactAmountIn(
...
+	// - the route is of length 2
+	// - route 1 and route 2 don't trade via the same pool
+	// - route 1 contains uosmo
+	// - both route 1 and route 2 are incentivized pools
+	// if all of the above is true, then we collect the additive and max fee between the two pools to later calculate the following:
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:31]:
```suggestion
	// - both route 1 and route 2 are incentivized pools
	//
	// If all of the above is true, then we collect the additive and max fee between the
	// two pools to later calculate the following:
```

File: x/gamm/keeper/multihop.go [hunk 5]
```diff
@@ -35,7 +53,14 @@ func (k Keeper) MultihopSwapExactAmountIn(
 			return sdk.Int{}, poolErr
 		}
 
-		swapFee := pool.GetSwapFee(ctx).Mul(swapFeeMultiplier)
+		swapFee := pool.GetSwapFee(ctx)
+
+		// if we determined the route is an osmo multi-hop and both routes are incentivized,
+		// we modify the swap fee accordingly
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:59]:
```suggestion
		// If we determined the route is an osmo multi-hop and both routes are incentivized,
		// we modify the swap fee accordingly.
```

File: x/gamm/keeper/multihop.go [hunk 6]
```diff
@@ -92,7 +140,10 @@ func (k Keeper) MultihopSwapExactAmountOut(
 		if poolErr != nil {
 			return sdk.Int{}, poolErr
 		}
-		swapFee := pool.GetSwapFee(ctx).Mul(swapFeeMultiplier)
+		swapFee := pool.GetSwapFee(ctx)
+		if isMultiHopRouted {
+			swapFee = routeSwapFee.Mul((swapFee.Quo(sumOfSwapFees)))
+		}
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:146]:
```suggestion

		swapFee := pool.GetSwapFee(ctx)
		if isMultiHopRouted {
			swapFee = routeSwapFee.Mul((swapFee.Quo(sumOfSwapFees)))
		}

```

File: x/gamm/keeper/multihop.go [hunk 7]
```diff
@@ -109,6 +160,53 @@ func (k Keeper) MultihopSwapExactAmountOut(
 	return tokenInAmount, nil
 }
 
+func (k Keeper) isOsmoRoutedMultihop(ctx sdk.Context,
+	route types.MultihopRoute,
+	inDenom string, outDenom string) (isRouted bool) {
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:165]:
```suggestion
func (k Keeper) isOsmoRoutedMultihop(ctx sdk.Context, route types.MultihopRoute, inDenom, outDenom string) (isRouted bool) {
```

File: x/gamm/keeper/multihop.go [hunk 8]
```diff
@@ -109,6 +160,53 @@ func (k Keeper) MultihopSwapExactAmountOut(
...
+	for _, poolId := range route.PoolIds() {
+		pool, poolErr := k.getPoolForSwap(ctx, poolId)
+		if poolErr != nil {
+			return sdk.Dec{}, sdk.Dec{}, poolErr
+		}
+		swapFee := pool.GetSwapFee(ctx)
+		additiveSwapFee = additiveSwapFee.Add(swapFee)
+		if swapFee.GT(maxSwapFee) {
+			maxSwapFee = swapFee
+		}
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:201]: You can just do `maxSwapFee = sdk.MaxDec(maxSwapFee, swapFee)`

File: x/gamm/keeper/multihop.go [hunk 9]
```diff
@@ -109,6 +160,53 @@ func (k Keeper) MultihopSwapExactAmountOut(
...
+	}
+	averageSwapFee := additiveSwapFee.QuoInt64(2)
+	if averageSwapFee.GT(maxSwapFee) {
```
Comment (alexanderbez) [inline x/gamm/keeper/multihop.go:204]: ditto

**Review by alexanderbez — APPROVED** [review]
LGTM!

**Pull request merged**